File: mobile/app/src/main/java/org/syncmob/mobile/net/Discovery.kt
package org.syncmob.mobile.net

import android.content.Context
import android.net.wifi.WifiManager
import android.util.Log
import org.syncmob.mobile.proto.Beacon
import org.syncmob.mobile.proto.Proto
import org.syncmob.mobile.security.Fingerprints
import org.syncmob.mobile.util.SafeFiles
import java.net.DatagramPacket
import java.net.Inet4Address
import java.net.InetAddress
import java.net.InetSocketAddress
import java.net.MulticastSocket
import java.net.NetworkInterface
import java.net.SocketTimeoutException
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicBoolean
import kotlin.concurrent.thread

/**
 * LAN presence over UDP broadcast and multicast.
 *
 * Beacons are unauthenticated **by design**: they carry only public information
 * and are treated purely as a hint about where a device might be reachable.
 * Anyone can forge one, so a beacon never grants trust — it only populates the
 * "nearby" list, and every entry still has to survive the Noise handshake and
 * the trust store.
 */
class Discovery(
    private val context: Context,
    private val ownPublicKey: ByteArray,
    private val onPeer: (DiscoveredPeer) -> Unit,
) {
    data class DiscoveredPeer(
        val deviceId: String,
        val publicKey: ByteArray,
        val name: String,
        val platform: String,
        val address: InetSocketAddress,
    ) {
        override fun equals(other: Any?): Boolean =
            other is DiscoveredPeer && deviceId == other.deviceId

        override fun hashCode(): Int = deviceId.hashCode()
    }

    @Volatile
    var announcedName: String = "Android"

    @Volatile
    var announcedPort: Int = Proto.DEFAULT_TCP_PORT

    @Volatile
    var announcing: Boolean = true

    private val running = AtomicBoolean(false)
    private var socket: MulticastSocket? = null
    private var multicastLock: WifiManager.MulticastLock? = null
    private val rateLimiter = RateLimiter()

    fun start() {
        if (!running.compareAndSet(false, true)) return
        try {
            acquireMulticastLock()
            val s = MulticastSocket(null).apply {
                reuseAddress = true
                bind(InetSocketAddress(Proto.DISCOVERY_PORT))
                soTimeout = 500
                broadcast = true
                timeToLive = 1
            }
            socket = s
            joinMulticast(s)
            thread(name = "syncmob-discovery-rx", isDaemon = true) { receiveLoop(s) }
            thread(name = "syncmob-discovery-tx", isDaemon = true) { announceLoop(s) }
        } catch (e: Exception) {
            Log.w(TAG, "discovery unavailable: ${e.message}")
            running.set(false)
            releaseMulticastLock()
        }
    }

    fun stop() {
        running.set(false)
        runCatching { socket?.close() }
        socket = null
        releaseMulticastLock()
    }

    private fun acquireMulticastLock() {
        val wifi = context.applicationContext.getSystemService(Context.WIFI_SERVICE) as? WifiManager
            ?: return
        multicastLock = wifi.createMulticastLock("syncmob").apply {
            setReferenceCounted(false)
            runCatching { acquire() }
        }
    }

    private fun releaseMulticastLock() {
        runCatching { multicastLock?.takeIf { it.isHeld }?.release() }
        multicastLock = null
    }

    private fun joinMulticast(s: MulticastSocket) {
        val group = InetSocketAddress(InetAddress.getByName(Proto.MULTICAST_GROUP), Proto.DISCOVERY_PORT)
        var joined = false
        for (nif in usableInterfaces()) {
            runCatching {
                s.joinGroup(group, nif)
                joined = true
            }
        }
        if (!joined) {
            runCatching { s.joinGroup(InetAddress.getByName(Proto.MULTICAST_GROUP)) }
        }
    }

    private fun receiveLoop(s: MulticastSocket) {
        val buf = ByteArray(Proto.MAX_BEACON_BYTES + 1)
        while (running.get()) {
            val packet = DatagramPacket(buf, buf.size)
            try {
                s.receive(packet)
            } catch (e: SocketTimeoutException) {
                continue
            } catch (e: Exception) {
                if (running.get()) Log.d(TAG, "receive error: ${e.message}")
                continue
            }
            val from = packet.address ?: continue
            if (packet.length > Proto.MAX_BEACON_BYTES) continue
            if (!SafeFiles.isLanAddress(from)) continue
            if (!rateLimiter.allow(from.hostAddress ?: continue)) continue

            val beacon = Beacon.parse(packet.data, packet.length) ?: continue
            // Ignore our own beacons echoing back off the network.
            if (beacon.publicKey.contentEquals(ownPublicKey)) continue

            onPeer(
                DiscoveredPeer(
                    deviceId = Fingerprints.deviceId(beacon.publicKey),
                    publicKey = beacon.publicKey,
                    name = beacon.name,
                    platform = beacon.platform,
                    // The address always comes from the UDP source, never from
                    // the payload: a beacon cannot point us at a third party.
                    address = InetSocketAddress(from, beacon.port),
                ),
            )
        }
    }

    private fun announceLoop(s: MulticastSocket) {
        val group = InetAddress.getByName(Proto.MULTICAST_GROUP)
        while (running.get()) {
            if (announcing) {
                val payload = Beacon(ownPublicKey, announcedName, announcedPort, "android").encode()
                if (payload.size <= Proto.MAX_BEACON_BYTES) {
                    val targets = mutableListOf<InetAddress>(group)
                    targets += broadcastAddresses()
                    for (t in targets) {
                        runCatching {
                            s.send(DatagramPacket(payload, payload.size, t, Proto.DISCOVERY_PORT))
                        }
                    }
                }
            }
            sleepInterruptible(ANNOUNCE_INTERVAL_MS)
        }
    }

    private fun sleepInterruptible(totalMs: Long) {
        var left = totalMs
        while (left > 0 && running.get()) {
            val step = minOf(200L, left)
            Thread.sleep(step)
            left -= step
        }
    }

    private fun usableInterfaces(): List<NetworkInterface> = runCatching {
        NetworkInterface.getNetworkInterfaces().toList()
            .filter { it.isUp && !it.isLoopback && it.supportsMulticast() }
    }.getOrDefault(emptyList())

    /** Per-interface broadcast addresses, plus the global one. */
    private fun broadcastAddresses(): List<InetAddress> {
        val out = mutableListOf<InetAddress>()
        runCatching { out += InetAddress.getByName("255.255.255.255") }
        for (nif in usableInterfaces()) {
            for (ia in nif.interfaceAddresses) {
                val b = ia.broadcast ?: continue
                if (b is Inet4Address) out += b
            }
        }
        return out.distinct()
    }

    /** Caps how many beacons a single source can push into the peer list. */
    private class RateLimiter {
        private val seen = ConcurrentHashMap<String, LongArray>()

        fun allow(host: String): Boolean {
            val now = System.currentTimeMillis()
            if (seen.size > 1024) {
                seen.entries.removeIf { now - it.value[0] > 5_000 }
                if (seen.size > 1024) return false
            }
            val entry = seen.getOrPut(host) { longArrayOf(now, 0) }
            synchronized(entry) {
                if (now - entry[0] >= 1_000) {
                    entry[0] = now
                    entry[1] = 0
                }
                entry[1]++
                return entry[1] <= RATE_PER_SOURCE
            }
        }
    }

    companion object {
        private const val TAG = "SyncMob/Discovery"
        const val ANNOUNCE_INTERVAL_MS = 3_000L
        /** A peer leaves the list after this long without a beacon. */
        const val PEER_TTL_MS = 15_000L
        const val MAX_PEERS = 128
        private const val RATE_PER_SOURCE = 4
    }
}

File: desktop/src/net/discovery.rs
//! LAN presence: UDP broadcast + multicast beacons.
//!
//! Beacons are **unauthenticated by design**. They carry only public
//! information (a public key, a display name, a port) and are treated purely as
//! a hint about where a device might be reachable. Anyone can forge one, so a
//! beacon never grants trust: it only populates the "nearby devices" list, and
//! every entry still has to survive the Noise handshake and the trust store.

use std::collections::HashMap;
use std::net::{IpAddr, Ipv4Addr, SocketAddr, SocketAddrV4, UdpSocket};
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::{Arc, Mutex};
use std::time::{Duration, Instant};

use base64::{engine::general_purpose::STANDARD as B64, Engine as _};
use socket2::{Domain, Protocol, Socket, Type};

use crate::proto::{Beacon, BEACON_MAGIC, DISCOVERY_PORT, MAX_BEACON_BYTES, MULTICAST_GROUP};
use crate::security::{device_id, DeviceId};
use crate::util::{clamp_str, is_lan_addr};

/// How often we announce ourselves.
pub const ANNOUNCE_INTERVAL: Duration = Duration::from_secs(3);
/// A peer disappears from the list after this long without a beacon.
pub const PEER_TTL: Duration = Duration::from_secs(15);
/// Upper bound on the peer table, so a flood of forged beacons cannot exhaust
/// memory or make the UI unusable.
pub const MAX_PEERS: usize = 128;
/// Beacons accepted per source address per second.
const RATE_PER_SOURCE: u32 = 4;

#[derive(Debug, Clone)]
pub struct DiscoveredPeer {
    pub device_id: DeviceId,
    /// Claimed X25519 static public key. Verified only by the handshake.
    pub public_key: Vec<u8>,
    /// Claimed display name. Untrusted; already clamped.
    pub name: String,
    pub platform: String,
    pub addr: SocketAddr,
}

/// What we announce. Behind a mutex so the UI can rename the device live.
#[derive(Debug, Clone)]
pub struct BeaconInfo {
    pub public_key: Vec<u8>,
    pub name: String,
    pub port: u16,
    pub platform: String,
    pub announce: bool,
}

pub struct Discovery {
    running: Arc<AtomicBool>,
    info: Arc<Mutex<BeaconInfo>>,
}

impl Discovery {
    /// Start the announce and listen loops. `on_peer` is called from the
    /// listener thread for every beacon that passes validation.
    pub fn start<F>(info: BeaconInfo, on_peer: F) -> std::io::Result<Self>
    where
        F: Fn(DiscoveredPeer) + Send + 'static,
    {
        let running = Arc::new(AtomicBool::new(true));
        let info = Arc::new(Mutex::new(info));

        let socket = bind_discovery_socket()?;
        join_multicast(&socket);
        socket.set_broadcast(true)?;
        socket.set_read_timeout(Some(Duration::from_millis(500)))?;

        let send_sock = socket.try_clone()?;
        let own_pk = info.lock().expect("info").public_key.clone();

        // Listener.
        {
            let running = running.clone();
            std::thread::Builder::new()
                .name("syncmob-discovery-rx".into())
                .spawn(move || {
                    let mut limiter = RateLimiter::new();
                    let mut buf = [0u8; MAX_BEACON_BYTES + 1];
                    while running.load(Ordering::Relaxed) {
                        let (n, src) = match socket.recv_from(&mut buf) {
                            Ok(v) => v,
                            Err(_) => continue, // timeout or transient error
                        };
                        if n > MAX_BEACON_BYTES || !limiter.allow(src.ip()) {
                            continue;
                        }
                        if let Some(peer) = validate_beacon(&buf[..n], src, &own_pk) {
                            on_peer(peer);
                        }
                    }
                })?;
        }

        // Announcer.
        {
            let running = running.clone();
            let info = info.clone();
            std::thread::Builder::new()
                .name("syncmob-discovery-tx".into())
                .spawn(move || {
                    while running.load(Ordering::Relaxed) {
                        let payload = {
                            let i = info.lock().expect("info");
                            if i.announce {
                                encode_beacon(&i)
                            } else {
                                None
                            }
                        };
                        if let Some(payload) = payload {
                            let targets = [
                                SocketAddr::from((Ipv4Addr::BROADCAST, DISCOVERY_PORT)),
                                SocketAddr::V4(SocketAddrV4::new(MULTICAST_GROUP, DISCOVERY_PORT)),
                            ];
                            for t in targets {
                                let _ = send_sock.send_to(&payload, t);
                            }
                            for bcast in interface_broadcasts() {
                                let _ = send_sock
                                    .send_to(&payload, SocketAddr::from((bcast, DISCOVERY_PORT)));
                            }
                        }
                        sleep_interruptible(&running, ANNOUNCE_INTERVAL);
                    }
                })?;
        }

        Ok(Self { running, info })
    }

    /// Change the announced name / port / platform at runtime.
    pub fn update(&self, f: impl FnOnce(&mut BeaconInfo)) {
        if let Ok(mut i) = self.info.lock() {
            f(&mut i);
        }
    }

    pub fn stop(&self) {
        self.running.store(false, Ordering::Relaxed);
    }
}

impl Drop for Discovery {
    fn drop(&mut self) {
        self.stop();
    }
}

fn sleep_interruptible(running: &AtomicBool, total: Duration) {
    let step = Duration::from_millis(200);
    let mut left = total;
    while left > Duration::ZERO && running.load(Ordering::Relaxed) {
        let d = step.min(left);
        std::thread::sleep(d);
        left -= d;
    }
}

fn bind_discovery_socket() -> std::io::Result<UdpSocket> {
    let socket = Socket::new(Domain::IPV4, Type::DGRAM, Some(Protocol::UDP))?;
    // Several SyncMob instances (or other multicast users) may share the port.
    socket.set_reuse_address(true)?;
    #[cfg(unix)]
    socket.set_reuse_port(true)?;
    socket.bind(&SocketAddr::from((Ipv4Addr::UNSPECIFIED, DISCOVERY_PORT)).into())?;
    Ok(socket.into())
}

fn join_multicast(socket: &UdpSocket) {
    // Join on every usable IPv4 interface; failures are non-fatal because
    // broadcast still works.
    let mut joined_any = false;
    if let Ok(ifaces) = if_addrs::get_if_addrs() {
        for iface in ifaces {
            if iface.is_loopback() {
                continue;
            }
            if let IpAddr::V4(v4) = iface.ip() {
                if socket.join_multicast_v4(&MULTICAST_GROUP, &v4).is_ok() {
                    joined_any = true;
                }
            }
        }
    }
    if !joined_any {
        let _ = socket.join_multicast_v4(&MULTICAST_GROUP, &Ipv4Addr::UNSPECIFIED);
    }
}

/// Per-interface broadcast addresses, so we also reach subnets where the
/// global 255.255.255.255 broadcast is filtered.
fn interface_broadcasts() -> Vec<Ipv4Addr> {
    let mut out = Vec::new();
    if let Ok(ifaces) = if_addrs::get_if_addrs() {
        for iface in ifaces {
            if iface.is_loopback() {
                continue;
            }
            if let if_addrs::IfAddr::V4(v4) = iface.addr {
                if let Some(b) = v4.broadcast {
                    out.push(b);
                }
            }
        }
    }
    out.sort();
    out.dedup();
    out
}

fn encode_beacon(info: &BeaconInfo) -> Option<Vec<u8>> {
    let b = Beacon {
        m: BEACON_MAGIC.to_string(),
        v: 1,
        pk: B64.encode(&info.public_key),
        name: clamp_str(&info.name, 32),
        port: info.port,
        platform: info.platform.clone(),
    };
    let raw = serde_json::to_vec(&b).ok()?;
    (raw.len() <= MAX_BEACON_BYTES).then_some(raw)
}

/// Parse and sanity-check a beacon. Pure function, so every rejection rule is
/// directly testable.
pub fn validate_beacon(raw: &[u8], src: SocketAddr, own_pk: &[u8]) -> Option<DiscoveredPeer> {
    if raw.len() > MAX_BEACON_BYTES {
        return None;
    }
    // Only ever talk to the local network.
    if !is_lan_addr(src.ip()) {
        return None;
    }
    let b: Beacon = serde_json::from_slice(raw).ok()?;
    if b.m != BEACON_MAGIC || b.v != 1 || b.port == 0 {
        return None;
    }
    let pk = B64.decode(&b.pk).ok()?;
    if pk.len() != 32 {
        return None;
    }
    // Ignore our own beacons coming back from the network.
    if pk == own_pk {
        return None;
    }
    Some(DiscoveredPeer {
        device_id: device_id(&pk),
        public_key: pk,
        name: clamp_str(&b.name, 32),
        platform: clamp_str(&b.platform, 16),
        addr: SocketAddr::new(src.ip(), b.port),
    })
}

/// Simple per-source token bucket.
struct RateLimiter {
    seen: HashMap<IpAddr, (Instant, u32)>,
}

impl RateLimiter {
    fn new() -> Self {
        Self {
            seen: HashMap::new(),
        }
    }
    fn allow(&mut self, ip: IpAddr) -> bool {
        let now = Instant::now();
        // Keep the map from growing without bound under a spoofed-source flood.
        if self.seen.len() > 1024 {
            self.seen
                .retain(|_, (t, _)| now.duration_since(*t) < Duration::from_secs(5));
            if self.seen.len() > 1024 {
                return false;
            }
        }
        let e = self.seen.entry(ip).or_insert((now, 0));
        if now.duration_since(e.0) >= Duration::from_secs(1) {
            *e = (now, 0);
        }
        e.1 += 1;
        e.1 <= RATE_PER_SOURCE
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn beacon_bytes(pk: &[u8], name: &str, port: u16) -> Vec<u8> {
        serde_json::to_vec(&Beacon {
            m: BEACON_MAGIC.into(),
            v: 1,
            pk: B64.encode(pk),
            name: name.into(),
            port,
            platform: "test".into(),
        })
        .unwrap()
    }

    fn lan_src() -> SocketAddr {
        "192.168.1.50:45820".parse().unwrap()
    }

    #[test]
    fn accepts_a_well_formed_beacon() {
        let raw = beacon_bytes(&[9u8; 32], "Phone", 45821);
        let p = validate_beacon(&raw, lan_src(), &[1u8; 32]).unwrap();
        assert_eq!(p.name, "Phone");
        assert_eq!(p.public_key, vec![9u8; 32]);
        assert_eq!(p.addr.port(), 45821);
        // The address always comes from the UDP source, never from the payload.
        assert_eq!(p.addr.ip(), lan_src().ip());
        assert_eq!(p.device_id, device_id(&[9u8; 32]));
    }

    #[test]
    fn rejects_non_lan_sources() {
        let raw = beacon_bytes(&[9u8; 32], "Phone", 45821);
        assert!(validate_beacon(&raw, "8.8.8.8:45820".parse().unwrap(), &[1u8; 32]).is_none());
    }

    #[test]
    fn rejects_own_beacon() {
        let raw = beacon_bytes(&[1u8; 32], "Me", 45821);
        assert!(validate_beacon(&raw, lan_src(), &[1u8; 32]).is_none());
    }

    #[test]
    fn rejects_malformed_payloads() {
        assert!(validate_beacon(b"not json", lan_src(), &[1u8; 32]).is_none());
        assert!(validate_beacon(&[], lan_src(), &[1u8; 32]).is_none());
        assert!(
            validate_beacon(&vec![b'{'; MAX_BEACON_BYTES + 1], lan_src(), &[1u8; 32]).is_none()
        );

        // Wrong magic / version / port / key length.
        let mut b: Beacon = serde_json::from_slice(&beacon_bytes(&[9u8; 32], "x", 1)).unwrap();
        b.m = "OTHER".into();
        assert!(validate_beacon(&serde_json::to_vec(&b).unwrap(), lan_src(), &[1u8; 32]).is_none());

        let raw = beacon_bytes(&[9u8; 32], "x", 0);
        assert!(validate_beacon(&raw, lan_src(), &[1u8; 32]).is_none());

        let raw = beacon_bytes(&[9u8; 16], "x", 45821);
        assert!(validate_beacon(&raw, lan_src(), &[1u8; 32]).is_none());
    }

    #[test]
    fn long_names_are_clamped() {
        let raw = beacon_bytes(&[9u8; 32], &"n".repeat(200), 45821);
        let p = validate_beacon(&raw, lan_src(), &[1u8; 32]).unwrap();
        assert!(p.name.chars().count() <= 33);
    }

    #[test]
    fn rate_limiter_caps_a_flood() {
        let mut l = RateLimiter::new();
        let ip: IpAddr = "192.168.1.9".parse().unwrap();
        let allowed = (0..50).filter(|_| l.allow(ip)).count();
        assert_eq!(allowed, RATE_PER_SOURCE as usize);
        // A different source is unaffected.
        assert!(l.allow("192.168.1.10".parse().unwrap()));
    }

    #[test]
    fn our_own_beacon_encodes_within_the_limit() {
        let info = BeaconInfo {
            public_key: vec![7u8; 32],
            name: "x".repeat(64),
            port: 45821,
            platform: "windows".into(),
            announce: true,
        };
        let raw = encode_beacon(&info).unwrap();
        assert!(raw.len() <= MAX_BEACON_BYTES);
        let p = validate_beacon(&raw, lan_src(), &[0u8; 32]).unwrap();
        assert_eq!(p.public_key, vec![7u8; 32]);
    }
}
